tests/unit/plugins/__snapshots__/test_writer.ambr

```diff
@@ -146,19 +141,255 @@
+  
+      linear_second = airflow_hidden_session_dependencies_module.get_linear_second(
+          linear_first
+      )
+  
+      pickle.dump(
+          linear_second,
+          open(
+              "/tmp/airflow_hidden_session_dependencies/variable_linear_second.pickle",
+              "wb",
+          ),
+      )
+  
+  
+  def task_linear_third():
+  
+      linear_first = pickle.load(
+          open(
+              "/tmp/airflow_hidden_session_dependencies/variable_linear_first.pickle",
+              "rb",
+          )
+      )
+  
+      linear_second = pickle.load(
+          open(
+              "/tmp/airflow_hidden_session_dependencies/variable_linear_second.pickle",
+              "rb",
+          )
+      )
+  
+      linear_third = airflow_hidden_session_dependencies_module.get_linear_third(
+          linear_first, linear_second
+      )
+  
+      pickle.dump(
+          linear_third,
+          open(
+              "/tmp/airflow_hidden_session_dependencies/variable_linear_third.pickle",
+              "wb",
+          ),
+      )
+  
+  
+  def task_setup():
+  
+      pickle_folder = pathlib.Path("/tmp").joinpath("airflow_hidden_session_dependencies")
+      if not pickle_folder.exists():
+          pickle_folder.mkdir()
+  
+  
+  def task_teardown():
+  
+      pickle_files = (
+          pathlib.Path("/tmp")
+          .joinpath("airflow_hidden_session_dependencies")
+          .glob("*.pickle")
+      )
+      for f in pickle_files:
+          f.unlink()
+  
+  
+  default_dag_args = {
+      "owner": "airflow",
+      "retries": 2,
+      "start_date": days_ago(1),
+  }
+  
+  with DAG(
+      dag_id="airflow_hidden_session_dependencies_dag",
+      schedule_interval="*/15 * * * *",
+      max_active_runs=1,
+      catchup=False,
+      default_args=default_dag_args,
+  ) as dag:
+  
+      a = PythonOperator(
+          task_id="a_task",
+          python_callable=task_a,
+      )
+  
+      linear_first_for_artifact_linear_second_and_downstream = PythonOperator(
+          task_id="linear_first_for_artifact_linear_second_and_downstream_task",
+          python_callable=task_linear_first_for_artifact_linear_second_and_downstream,
+      )
+  
+      linear_second = PythonOperator(
+          task_id="linear_second_task",
+          python_callable=task_linear_second,
+      )
+  
+      linear_third = PythonOperator(
+          task_id="linear_third_task",
+          python_callable=task_linear_third,
+      )
+  
+      setup = PythonOperator(
+          task_id="setup_task",
+          python_callable=task_setup,
+      )
+  
+      teardown = PythonOperator(
+          task_id="teardown_task",
+          python_callable=task_teardown,
+      )
+  
+      a >> linear_first_for_artifact_linear_second_and_downstream
+  
+      linear_first_for_artifact_linear_second_and_downstream >> linear_second
+  
+      linear_first_for_artifact_linear_second_and_downstream >> linear_third
+  
+      linear_second >> linear_third
```

Comment:
This over-specified dependency looks a bit awkward but should be fine in execution, right?

cc @mingjerli

Reply:
This is fine in execution. 

The reason why this thing looks awkward is within a session, the algorithm gives us a minimal edge representation of the DAG but it is no longer the case when we handle user-specified inter-session dependencies. 

The benefit of minimal edge representation is if you break any edge, the DAG must break into to two graphs; with any extra edges, it is no longer the case. In practice, if I want to manually break one pipeline into two pipelines, with minimal edge representation, I only need to find out the exact edge I want to remove; this is will make debugging dependencies related issues much easier than other representation. I also believe the minimal edge representation is unique, which is more elegant in my opinion.

Reply:
Okay, so this is another instance of correct but less optimal, which we decided to be okay as per @andycui97's discussion with DMX.